File: agent/lib/core/protocol.dart
/// Mirrors the server's signalling protocol (see server/src/protocol.ts).
/// Kept in sync manually with the agent app.
library;

/// Message envelope.
class WsEnvelope {
  final String id;
  final String type;
  final dynamic payload;
  final int ts;

  const WsEnvelope({required this.id, required this.type, required this.payload, required this.ts});

  Map<String, dynamic> toJson() => {'id': id, 'type': type, 'payload': payload, 'ts': ts};

  factory WsEnvelope.fromJson(Map<String, dynamic> json) => WsEnvelope(
        id: json['id'] as String,
        type: json['type'] as String,
        payload: json['payload'],
        ts: (json['ts'] as num?)?.toInt() ?? 0,
      );
}

/// Helper to build envelopes.
class Protocol {
  static String _uid() => DateTime.now().microsecondsSinceEpoch.toString() + (DateTime.now().millisecond).toString();

  static WsEnvelope auth({
    required String deviceId,
    required String role,
    required String token,
    String? name,
    String? model,
    String? androidVersion,
    String? appVersion,
    Map<String, dynamic>? rejoin,
  }) =>
      WsEnvelope(
        id: _uid(),
        type: 'AUTH',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {
          'deviceId': deviceId,
          'role': role,
          'token': token,
          'name': ?name,
          'model': ?model,
          'androidVersion': ?androidVersion,
          'appVersion': ?appVersion,
          'rejoin': ?rejoin,
        },
      );

  static WsEnvelope pairRequest({required String deviceId}) => WsEnvelope(
        id: _uid(),
        type: 'PAIR_REQUEST',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'deviceId': deviceId},
      );

  static WsEnvelope pairApproved({required String pairId}) => WsEnvelope(
        id: _uid(),
        type: 'PAIR_APPROVED',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'pairId': pairId},
      );

  static WsEnvelope pairDenied({required String pairId}) => WsEnvelope(
        id: _uid(),
        type: 'PAIR_DENIED',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'pairId': pairId},
      );

  static WsEnvelope sessionAccept({required String sessionId, required List<String> permissions}) => WsEnvelope(
        id: _uid(),
        type: 'SESSION_ACCEPT',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'permissions': permissions},
      );

  static WsEnvelope sessionDeny({required String sessionId, String? reason}) => WsEnvelope(
        id: _uid(),
        type: 'SESSION_DENY',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'reason': ?reason},
      );

  static WsEnvelope sessionRejoin({required String sessionId, required String sessionToken}) => WsEnvelope(
        id: _uid(),
        type: 'SESSION_REJOIN',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'sessionToken': sessionToken},
      );

  static WsEnvelope offer({required String sessionId, required String sdp}) => WsEnvelope(
        id: _uid(),
        type: 'OFFER',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'sdp': sdp},
      );

  static WsEnvelope answer({required String sessionId, required String sdp}) => WsEnvelope(
        id: _uid(),
        type: 'ANSWER',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'sdp': sdp},
      );

  static WsEnvelope iceCandidate({required String sessionId, required Map<String, dynamic> candidate}) => WsEnvelope(
        id: _uid(),
        type: 'ICE_CANDIDATE',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'candidate': candidate},
      );

  static WsEnvelope touch({
    required String sessionId,
    required String eventType,
    required double x,
    required double y,
    required int screenWidth,
    required int screenHeight,
    int? action,
  }) =>
      WsEnvelope(
        id: _uid(),
        type: 'TOUCH_EVENT',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {
          'sessionId': sessionId,
          'eventType': eventType,
          'x': x,
          'y': y,
          'timestamp': DateTime.now().millisecondsSinceEpoch,
          'screenWidth': screenWidth,
          'screenHeight': screenHeight,
          'action': ?action,
        },
      );

  static WsEnvelope control({required String sessionId, required String kind, dynamic value}) => WsEnvelope(
        id: _uid(),
        type: 'CONTROL_EVENT',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'kind': kind, 'value': ?value},
      );

  static WsEnvelope fileRequest({
    required String sessionId,
    required String transferId,
    required String fileName,
    required int fileSize,
    required String mimeType,
    String? sha256,
    int? chunkSize,
  }) =>
      WsEnvelope(
        id: _uid(),
        type: 'FILE_REQUEST',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {
          'sessionId': sessionId,
          'transferId': transferId,
          'fileName': fileName,
          'fileSize': fileSize,
          'mimeType': mimeType,
          'sha256': ?sha256,
          'chunkSize': ?chunkSize,
        },
      );

  static WsEnvelope fileAccept({required String sessionId, required String transferId}) => WsEnvelope(
        id: _uid(),
        type: 'FILE_ACCEPT',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'transferId': transferId},
      );

  static WsEnvelope fileDecline({required String sessionId, required String transferId, String? reason}) => WsEnvelope(
        id: _uid(),
        type: 'FILE_DECLINE',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'transferId': transferId, 'reason': ?reason},
      );

  static WsEnvelope fileProgress({required String sessionId, required String transferId, required int transferred}) => WsEnvelope(
        id: _uid(),
        type: 'FILE_PROGRESS',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'transferId': transferId, 'transferred': transferred},
      );

  static WsEnvelope fileComplete({required String sessionId, required String transferId, String? sha256}) => WsEnvelope(
        id: _uid(),
        type: 'FILE_COMPLETE',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'transferId': transferId, 'sha256': ?sha256},
      );

  static WsEnvelope clipboard({required String sessionId, required String direction, required String text}) => WsEnvelope(
        id: _uid(),
        type: 'CLIPBOARD_EVENT',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'direction': direction, 'text': text},
      );

  static WsEnvelope permissionUpdate({required String sessionId, required String permission, required bool granted}) => WsEnvelope(
        id: _uid(),
        type: 'PERMISSION_UPDATE',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'permission': permission, 'granted': granted},
      );

  static WsEnvelope disconnect({required String sessionId, String? reason}) => WsEnvelope(
        id: _uid(),
        type: 'DISCONNECT',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'sessionId': sessionId, 'reason': ?reason},
      );

  static WsEnvelope ping() => WsEnvelope(
        id: _uid(),
        type: 'PING',
        ts: DateTime.now().millisecondsSinceEpoch,
        payload: {'ts': DateTime.now().millisecondsSinceEpoch},
      );
}

/// Permission identifiers shared with the server.
class Permissions {
  static const List<String> all = [
    'SCREEN',
    'TOUCH',
    'CAMERA',
    'MICROPHONE',
    'FILES',
    'CLIPBOARD',
    'DEVICE_INFO',
  ];

  static const Map<String, String> labels = {
    'SCREEN': 'Screen',
    'TOUCH': 'Touch Control',
    'CAMERA': 'Camera',
    'MICROPHONE': 'Microphone',
    'FILES': 'File Transfer',
    'CLIPBOARD': 'Clipboard',
    'DEVICE_INFO': 'Device Info',
  };

  static String label(String p) => labels[p] ?? p;
}

File: server/src/protocol.ts
import { z } from 'zod';
import { Envelope, SessionMessageType } from './types.js';

export const permissionSchema = z.enum([
  'SCREEN',
  'TOUCH',
  'CAMERA',
  'MICROPHONE',
  'FILES',
  'CLIPBOARD',
  'DEVICE_INFO',
]);

export const permissionListSchema = z.array(permissionSchema).max(7);

const idString = z.string().min(1).max(128);
const tokenString = z.string().min(1).max(2048);
const reasonSchema = z.string().min(1).max(200);

export const messageSchemas: Record<SessionMessageType, z.ZodTypeAny> = {
  AUTH: z.object({
    deviceId: idString,
    role: z.enum(['controller', 'agent']),
    token: tokenString,
    name: z.string().max(64).optional(),
    model: z.string().max(64).optional(),
    androidVersion: z.string().max(32).optional(),
    appVersion: z.string().max(32).optional(),
    rejoin: z.object({ sessionId: idString, sessionToken: tokenString }).optional(),
  }),

  AUTH_OK: z.object({
    deviceId: idString,
    role: z.enum(['controller', 'agent']),
    onlineDevices: z.array(z.string()),
    activeSessionId: z.string().optional(),
  }),

  AUTH_ERROR: z.object({ code: idString, message: z.string() }).optional(),

  ERROR: z.object({ code: idString, message: z.string(), refId: idString.optional() }).optional(),

  HELLO: z.object({ status: z.string() }),

  // Agent requests a fresh pairing code
  PAIR_REQUEST: z.object({ deviceId: idString }),

  PAIR_RESPONSE: z.object({
    deviceId: idString,
    code: z.string(),
    expiresAt: z.number(),
    expiresIn: z.number(),
  }),

  // Server -> agent: someone wants to pair with this agent
  PAIR_APPROVAL: z.object({
    pairId: idString,
    controllerDeviceId: idString,
    controllerName: z.string(),
    requestedAt: z.number(),
  }),

  // Agent -> server: approve/deny pairing
  PAIR_APPROVED: z.object({ pairId: idString }),
  PAIR_DENIED: z.object({ pairId: idString }),

  // Server -> agent: a controller is requesting remote access
  SESSION_REQUEST: z.object({
    sessionId: idString,
    controllerDeviceId: idString,
    controllerName: z.string(),
    permissions: permissionListSchema,
    iceServers: z.array(z.unknown()).default([]),
    requestedAt: z.number(),
  }),

  // Agent -> server (accept/deny/reduce), then server -> controller
  SESSION_ACCEPT: z.object({
    sessionId: idString,
    permissions: permissionListSchema,
  }),
  SESSION_DENY: z.object({ sessionId: idString, reason: reasonSchema.optional() }),

  // Server -> controller: session established + session token for rejoin
  SESSION_ACCEPTED: z.object({
    sessionId: idString,
    permissions: permissionListSchema,
    sessionToken: tokenString,
    iceServers: z.array(z.unknown()).default([]),
  }),

  // Rejoin after network flip
  SESSION_REJOIN: z.object({ sessionId: idString, sessionToken: tokenString }),

  // Server -> both: session ended
  SESSION_ENDED: z.object({
    sessionId: idString,
    reason: reasonSchema,
    endedBy: z.enum(['controller', 'agent', 'server']),
  }),

  OFFER: z.object({ sessionId: idString, sdp: z.string().min(1) }),
  ANSWER: z.object({ sessionId: idString, sdp: z.string().min(1) }),
  ICE_CANDIDATE: z.object({
    sessionId: idString,
    candidate: z.unknown(),
  }),

  TOUCH_EVENT: z.object({
    sessionId: idString,
    eventType: z.enum(['touch_down', 'touch_move', 'touch_up', 'tap', 'long_press', 'swipe']),
    x: z.number().min(0).max(1),
    y: z.number().min(0).max(1),
    timestamp: z.number(),
    screenWidth: z.number().int().positive(),
    screenHeight: z.number().int().positive(),
    action: z.number().int().optional(),
  }),

  CONTROL_EVENT: z.object({
    sessionId: idString,
    kind: z.enum([
      'NAV_BACK',
      'NAV_HOME',
      'NAV_RECENTS',
      'TEXT_INPUT',
      'CAMERA_ON',
      'CAMERA_OFF',
      'MIC_ON',
      'MIC_OFF',
      'DEVICE_STATUS',
      'PERMISSION_GRANT',
      'PERMISSION_REVOKE',
    ]),
    value: z.union([z.string(), z.number(), z.boolean(), z.object({}).passthrough()]).optional(),
  }),

  FILE_REQUEST: z.object({
    sessionId: idString,
    transferId: idString,
    fileName: z.string().max(255),
    fileSize: z.number().int().nonnegative(),
    mimeType: z.string().max(128),
    sha256: z.string().max(64).optional(),
    chunkSize: z.number().int().positive().optional(),
  }),

  FILE_ACCEPT: z.object({ sessionId: idString, transferId: idString }),
  FILE_DECLINE: z.object({ sessionId: idString, transferId: idString, reason: reasonSchema.optional() }),

  FILE_PROGRESS: z.object({
    sessionId: idString,
    transferId: idString,
    transferred: z.number().int().nonnegative(),
  }),

  FILE_COMPLETE: z.object({
    sessionId: idString,
    transferId: idString,
    sha256: z.string().max(64).optional(),
  }),

  CLIPBOARD_EVENT: z.object({
    sessionId: idString,
    direction: z.enum(['controller_to_agent', 'agent_to_controller']),
    text: z.string().max(1048576),
  }),

  PERMISSION_UPDATE: z.object({
    sessionId: idString,
    permission: permissionSchema,
    granted: z.boolean(),
  }),

  DISCONNECT: z.object({ sessionId: idString, reason: reasonSchema.optional() }),

  PING: z.object({ ts: z.number() }),
  PONG: z.object({ ts: z.number() }),
};

export function validateMessage(input: unknown): { ok: true; value: Envelope } | { ok: false; error: string } {
  if (typeof input !== 'object' || input === null) return { ok: false, error: 'expected object' };
  const raw = input as Record<string, unknown>;
  const type = raw.type as SessionMessageType;
  const schema = messageSchemas[type];
  if (!schema) return { ok: false, error: `unknown message type: ${String(raw.type)}` };
  const id = typeof raw.id === 'string' ? raw.id : '';
  if (!id) return { ok: false, error: 'missing message id' };
  const ts = typeof raw.ts === 'number' ? raw.ts : Date.now();
  const parsed = schema.safeParse(raw.payload);
  if (!parsed.success) {
    return { ok: false, error: `invalid payload for ${type}: ${parsed.error.issues[0]?.message ?? 'validation failed'}` };
  }
  return { ok: true, value: { id, type, payload: parsed.data, ts } };
}